lib/spack/spack/cmd/install.py

```diff
@@ -489,43 +341,164 @@ def install(parser, args):
     if args.deprecated:
         spack.config.set("config:deprecated", True, scope="command_line")
 
-    spack.cmd.common.arguments.sanitize_reporter_options(args)
+    arguments.sanitize_reporter_options(args)
 
     def reporter_factory(specs):
         if args.log_format is None:
-            return None
+            return lang.nullcontext()
 
-        context_manager = spack.report.build_context_manager(
+        return spack.report.build_context_manager(
             reporter=args.reporter(), filename=report_filename(args, specs=specs), specs=specs
         )
-        return context_manager
 
     install_kwargs = install_kwargs_from_args(args)
 
-    if not args.spec and not args.specfiles:
-        # If there are no args but an active environment then install the packages from it.
-        install_all_specs_from_active_environment(
-            install_kwargs=install_kwargs,
-            only_concrete=args.only_concrete,
-            cli_test_arg=args.test,
-            reporter_factory=reporter_factory,
-        )
+    env = ev.active_environment()
+
+    if not env and not args.spec and not args.specfiles:
+        _die_require_env()
+
+    try:
+        if env:
+            install_with_active_env(env, args, install_kwargs, reporter_factory)
+        else:
+            install_without_active_env(args, install_kwargs, reporter_factory)
+    except spack.build_environment.InstallError as e:
+        if args.show_log_on_error:
+            _dump_log_on_error(e)
+        raise
+
+
+# Install inside environments
+
+
+def _filter_specs_in_env(env: ev.Environment, abstract_specs: List[spack.spec.Spec]):
+    return [
+        s
+        for s in traverse.traverse_nodes(env.concrete_roots(), key=lambda s: s.dag_hash())
+        if any(s.satisfies(t) for t in abstract_specs)
+    ]
```

Comment:
Can we add this to `Environment`? I think it would be good to have an API to install only the specs matching some abstract requirement (i.e. a partial installation of the environment). A first application might be in the bootstrap environment to distinguish `--dev` dependencies from runtime dependencies if we add them.